### ISSUE TITLE
fatal error: concurrent map read and map write

### ISSUE THREAD

**Issue by jebbench** [chartmuseum/chartmuseum#10]
When attempting to launch Chartmuseum we sometimes get the error `fatal error: concurrent map read and map write`.

Observed with both v0.1.0 and v0.2.0.

Running on Ubuntu 16.04 using Systemd with the below service configuration:
```
[Unit]
Description=Charmuseum service
After=network.target
After=systemd-user-sessions.service
After=network-online.target


[Service]
ExecStart=/opt/chartmuseum/chartmuseum --port=9001 --storage="amazon" --storage-amazon-bucket="my-bucket" --storage-amazon-prefix="" --storage-amazon-region="eu-west-1"

[Install]
WantedBy=multi-user.target
```

Log file: 
[chartmuseum.log](https://github.com/chartmuseum/chartmuseum/files/1382660/chartmuseum.log)

**Comment by cjauvin** [comment]
After a quick analysis (I am very new to this codebase so please bear with me), it seems to me that this concurrent call:

https://github.com/chartmuseum/chartmuseum/blob/master/pkg/chartmuseum/server.go#L226

which ultimately modify this map:

https://github.com/chartmuseum/chartmuseum/blob/master/pkg/repo/index.go#L67

which can be found in the Helm codebase:

https://github.com/kubernetes/helm/blob/master/pkg/repo/index.go#L81

might be the culprit.

**Comment by jdolitsky** [comment]
yikes. @cjauvin thanks for checking into this, I think you're right. That's the only piece of code with parallel execution..

@jebbench thanks for opening this. may I ask how many packages you have in storage?

**Comment by jdolitsky** [comment]
@jebbench also, I see at the end of the logs
`Starting ChartMuseum        {"port": 9001}`
does the service end up actually listening?

**Comment by jebbench** [comment]
> @jebbench also, I see at the end of the logs
> Starting ChartMuseum {"port": 9001}
> does the service end up actually listening? Or is this systemd retrying and succeeding?

This is me manually restarting it; sometimes takes a couple of attempts before it runs.

 > @jebbench thanks for opening this. may I ask how many packages you have in storag
Right now - 184; mostly different versions of the same chart (our CI server builds and publishes it); we do have plans for some sort of clean up script (e.g. delete anything ending -CI after 30 days) in the works.

**Comment by jdolitsky** [comment]
ok. on this surface this seems to be a symptom of having a lot of packages with the same name. will dive into this soon

**Comment by jdolitsky** [comment]
@jebbench Can you please try using the latest binary or Docker "latest" tag?

I am hoping that https://github.com/chartmuseum/chartmuseum/commit/6cc9c3fe4363f95f872c016a6003ce045e1947d2 has fixed this issue

**Comment by cjauvin** [comment]
@jdolitsky I'm not 100% sure, but I suspect that appending to the `loaded` slice as you do now is not goroutine-safe either.. Could it be done with a channel instead?

**Comment by jdolitsky** [comment]
Yea I think you're right. @cjauvin What would this look like with a channel?

Could we also construct the `loaded` slice with length `len(objects)`, then instead of appending, do `loaded[x] = chartVersion` ?

**Comment by cjauvin** [comment]
Very interesting question.. I agree that on the surface it would appear to
be safer, but I honestly don't know, and I suspect it might not be 100%
safe either for some subtle reason.. I'm looking into it right now.

**Comment by jdolitsky** [comment]
I think 00f7694b24191fb94373b23d45fb0cc2b6871c3f has fixed this issue

I've tried running this command with the `-race` flag:
```bash
go run -race cmd/chartmuseum/main.go --storage=local --storage-local-rootdir=./mirror --debug
```

On checkout of 6cc9c3fe4363f95f872c016a6003ce045e1947d2 I get the following error:
```
==================
WARNING: DATA RACE
Read at 0x00c420332b20 by goroutine 41:
  github.com/chartmuseum/chartmuseum/pkg/chartmuseum.(*Server).addIndexObjectsAsync.func1()
      --/src/github.com/chartmuseum/chartmuseum/pkg/chartmuseum/server.go:283 +0x171

Previous write at 0x00c420332b20 by goroutine 7:
  github.com/chartmuseum/chartmuseum/pkg/chartmuseum.(*Server).addIndexObjectsAsync.func1()
      --/src/github.com/chartmuseum/chartmuseum/pkg/chartmuseum/server.go:283 +0x214

Goroutine 41 (running) created at:
  github.com/chartmuseum/chartmuseum/pkg/chartmuseum.(*Server).addIndexObjectsAsync()
```

But running the same command on a checkout of 00f7694b24191fb94373b23d45fb0cc2b6871c3f (latest) prints no warnings.

I'm ready to close this issue once we have confirmation from @jebbench.
Thanks @cjauvin for extra eyes on this!

**Comment by cjauvin** [comment]
> Yea I think you're right. @cjauvin What would this look like with a channel?

I'm not suggesting you use it, but for the sake of increasing my understanding of Go channels, I wrote this, which I think would work:

https://github.com/chartmuseum/chartmuseum/compare/master...cjauvin:channel-instead-of-slice

**Comment by jdolitsky** [comment]
@jebbench The fix for this is now in v0.2.2. Please let us know if you encounter this error again. Thanks!

**Comment by jebbench** [comment]
I've updated and restarted the service a few times and it's restarted successfully each time.
